### ISSUE TITLE
.dockerignore not considered when computing cache keys

### ISSUE THREAD

**Issue by michaelhixson** [GoogleContainerTools/kaniko#871]
**Actual behavior**
Files excluded by `.dockerignore` are wrongly included when determining whether the cache can be used for a `COPY` build step.

**Expected behavior**
Modifications to files excluded by `.dockerignore` do not affect the cacheability of `COPY` steps.

**To Reproduce**

## Project files

**example/project/.dockerignore**
```
files/a.txt
```

**example/project/Dockerfile**
```
FROM alpine
COPY files files
CMD ["ls", "files"]
```

**example/project/files/a.txt**
```
hi
```

**example/project/files/b.txt**
```
bye
```

**example/docker_config.json**
```
{
  "auths": {
    REDACTED
  }
}
```

**example/docker-compose.yml** [&ast;]
```
version: "3.7"
services:
  foo:
    image: gcr.io/kaniko-project/executor:latest
    volumes:
      - ./project:/workspace
      - ./docker_config.json:/kaniko/.docker/config.json
    command:
      - --context=/workspace
      - --destination=REDACTED
      - --cache=true
```
[&ast;] *I would have used `docker run ...` instead of `docker-compose` except I couldn't figure out the Windows-compatible syntax that `docker run` wanted for mounting volumes.*

## Command line instructions

I do not include the full output of each command.  Instead I include my own comments about the output `### formatted like this`.

```
$ cd example

$ docker build project
### OK: docker won't use a cache for the "COPY files files" step
###     since it's the first build

$ docker build project
### OK: this time docker will use a cache for "COPY files files"

$ echo "hi2" >> project/files/a.txt
$ docker build project
### OK: docker will still use a cache for "COPY files files"
###     because files/a.txt was .dockerignored!

$ docker-compose up --build
### OK: kaniko won't use a cache for the "COPY files files" step
###     since it's the first build

$ docker-compose up --build
### OK: this time kaniko will use a cache for "COPY files files"

$ echo "hi3" >> project/files/a.txt
$ docker-compose up --build
### BAD: kaniko won't use a cache for "COPY files files"
###      but it should, since files/a.txt was dockerignored
```

**Additional Information**

```
$ docker images gcr.io/kaniko-project/executor --digests
REPOSITORY                       TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
gcr.io/kaniko-project/executor   latest              sha256:9c40a04cf1bc9d886f7f000e0b7fa5300c31c89e2ad001e97eeeecdce9f07a29   4afd732d4dfa        12 days ago         61.3MB
```

**My theory for the issue's cause**
While `copy.go`'s `ExecuteCommand` [takes exclusions from `.dockerignore` into account](https://github.com/GoogleContainerTools/kaniko/blob/56e048e044a64ac88d27c7c2c978ebaf28355af1/pkg/commands/copy.go#L91), the `FilesUsedFromContext` function from that same file [does not consider exclusions](https://github.com/GoogleContainerTools/kaniko/blob/56e048e044a64ac88d27c7c2c978ebaf28355af1/pkg/commands/copy.go#L114-L133), and that second function is what's used to [compute cache keys](https://github.com/GoogleContainerTools/kaniko/blob/9eb4a1c1ffd000fe4a318f145e7cd2ce54fcc576/pkg/executor/build.go#L151-L160).  Either `FilesUsedFromContext` or its caller in `build.go` should be changed to be aware of `.dockerignore`.
 
 **Triage Notes for the Maintainers**
 
 | **Description** | **Yes/No** |
 |----------------|---------------|
 | Please check if this a new feature you are proposing        | <ul><li>- [ ] </li></ul>|
 | Please check if the build works in docker but not in kaniko | <ul><li>- [x] </li></ul>| 
 | Please check if this error is seen when you use `--cache` flag | <ul><li>- [x] </li></ul>|
 | Please check if your dockerfile is a multistage dockerfile | <ul><li>- [ ] </li></ul>|

**Comment by michaelhixson** [comment]
Oops, I didn't see https://github.com/GoogleContainerTools/kaniko/pull/854 before filing this, but I bet that PR will fix this.

**Comment by tejal29** [comment]
This is closed via #924 and released in kaniko v0.17.0
